### PR TITLE
[bounty] Adminbuse race conversion/hypnosis machine!

### LINKED ISSUE #1985: [Code Bounty] Adminbus Felinid Event Machine
image
image
image

Currently this will be $10 paid via bounty source, paypal, or mainstream cryptocurrency.

Rationale:
I would like to see a machine that converts crew into felinids because doing it as an admin requires tedious use of VV per player converted. The players that were on that round had a fun time revolting against the captain, and it was hilarious to watch.

Deliverable:
A machine that converts a player inserted into a felinid (species). 
There should be an additional variant that brainwashes the victim with an objective of converting additional crew into felinids.
Any style of machine would work, from a machine styled after the malf AI's borg factory, or the medical machine that inserts implants.
It would also be nice if the machine could be configured as to what species is output, because then the opportunity for racewar can go beyond just felinids (lizards, moths, etc)

### PR DESCRIPTION
**Pull request by utilbird**
Bounty made by @bloons3 
Closes #1985

## About The Pull Request

Adds the "species conversion chamber" and a hypnosis variant. 

Being (forced) inside the chamber converts you into a pre-selected roundstart race after a short period of time. The hypnosis variant brainwashes victims with an objective to convert as many people as possible to their new race

## Why It's Good For The Game

https://www.youtube.com/watch?v=3oEWA7UglB4

## Changelog
:cl:TheFakeElon, bloons3
add: Adds an adminbuse species conversion chamber
/:cl:

### CODE REVIEW

File: code/game/machinery/race_converter.dm [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/obj/machiney/species_converter
+	name = "species conversion chamber"
+	desc = "Safely and efficiently converts the species of the subject, warrenty void if exposed to plasma."
```
Comment (bloons3) [inline code/game/machinery/race_converter.dm]: warranty?

File: code/game/machinery/race_converter.dm [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+				overlays += "[icon_state]_red"
+	else if(powered(AREA_USAGE_EQUIP))
+		overlays += "[icon_state]_red"
+	if(panel_open)
+		overlays += "[icon_state]_panel"
+
+/obj/machinery/species_converter/process()
+	if(!processing)
+		return
+	if(!is_operational() || !occupant || !iscarbon(occupant))
```
Comment (bloons3) [inline code/game/machinery/race_converter.dm:87]: hooray, it's a machine you don't get stuck inside when the power goes out!

**Comment by TeomanTheGreat** [comment]
Human supremacist conversion machine. Yes.

File: code/game/machinery/race_converter.dm [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+/obj/machinery/species_converter/New()
+	if(!length(possible_races))
+		for(var/datum/species/ST in subtypesof(/datum/species))
+			if(ST.check_roundstart_eligible())
+				possible_races += ST
+		possible_races = sortList(possible_races)
+	..()
+
+/obj/machinery/species_converter/AltClick(mob/user)
+	if(!user.canUseTopic(src, !issilicon(user)) || processing)
```
Comment (bloons3) [inline code/game/machinery/race_converter.dm]: non-admin usable machine?

**Review by bloons3 — COMMENTED** [review]
left comments

**Comment by ToasterBan** [comment]
felinid hypnosis chamber

**Comment by Naevii** [comment]
turn people into felinids O.o

**Comment by ghost** [comment]
make it able to change gender too

**Comment by PowerfulBacon** [comment]
you should definitely include subtypes for each race

File: code/game/machinery/race_converter.dm [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		var/mob/living/carbon/C = occupant
+		if(!is_species(C, desired_race))
+			processing = TRUE
+			soundloop.start()
+			update_icon()
+			set_light(2, 1, "#ff0000")
+		else
+			say("Occupant is already the desired race.")
+			playsound(src, 'sound/machines/buzz-sigh.ogg', 40, FALSE)
+			overlays += "[icon_state]_red"
```
Comment (park66665) [inline code/game/machinery/race_converter.dm]:
```suggestion
			open_machine()
```

File: code/game/machinery/race_converter.dm [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			if(occupant)
+				if(powered(AREA_USAGE_EQUIP))
+					overlays += "[icon_state]_stack"
+					overlays += "[icon_state]_yellow"
+			else
+				overlays += "[icon_state]_red"
+	else if(powered(AREA_USAGE_EQUIP))
+		overlays += "[icon_state]_red"
+	if(panel_open)
+		overlays += "[icon_state]_panel"
```
Comment (park66665) [inline code/game/machinery/race_converter.dm]:
```suggestion
/obj/machinery/species_converter/update_overlays()
	. = ..()
	if(!state_open)
		if(processing)
			. += "[icon_state]_door_on"
			. += "[icon_state]_stack"
			. += "[icon_state]_smoke"
			. += "[icon_state]_green"
		else
			. += "[icon_state]_door_off"
			if(occupant)
				if(powered(AREA_USAGE_EQUIP))
					. += "[icon_state]_stack"
					. += "[icon_state]_yellow"
			else
				. += "[icon_state]_red"
	else if(powered(AREA_USAGE_EQUIP))
		. += "[icon_state]_red"
	if(panel_open)
		. += "[icon_state]_panel"
```

**Review by park66665 — CHANGES_REQUESTED** [review]
Generally, code shouldn't touch `overlays` list directly.

File: code/game/machinery/race_converter.dm [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	if(panel_open)
+		to_chat(user, "<span class='warning'>You need to close the maintenance hatch first!</span>")
+		return
+	..()
+	playsound(src, 'sound/machines/click.ogg', 50)
+	if(occupant)
+		if(!iscarbon(occupant))
+			occupant.forceMove(drop_location())
+			occupant = null
+			return
```
Comment (park66665) [inline code/game/machinery/race_converter.dm]:
```suggestion
/obj/machinery/species_converter/can_be_occupant(atom/movable/am)
	return ishuman(am)

/obj/machinery/species_converter/close_machine(mob/user)
	if(panel_open)
		to_chat(user, "<span class='warning'>You need to close the maintenance hatch first!</span>")
		return
	..()
	playsound(src, 'sound/machines/click.ogg', 50)
	if(occupant)
```

`can_be_occupant` is called at `/obj/machinery/proc/close_machine`.

File: code/game/machinery/race_converter.dm [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	iterations = 0
+	playsound(src, 'sound/machines/click.ogg', 50)
+	if(processing)
+		stop()
+	..()
+
+/obj/machinery/species_converter/proc/stop()
+	processing = FALSE
+	soundloop.stop()
+	set_light(0, 0)
```
Comment (park66665) [inline code/game/machinery/race_converter.dm]:
```suggestion
/obj/machinery/species_converter/open_machine(mob/user)
	playsound(src, 'sound/machines/click.ogg', 50)
	if(processing)
		stop()
	..()

/obj/machinery/species_converter/proc/stop()
	processing = FALSE
	iterations = 0
	soundloop.stop()
	set_light(0, 0)
```

File: code/game/machinery/race_converter.dm [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+		if(brainwash)
+			to_chat(C, "<span class='userdanger'>A new compulsion fills your mind... you feel forced to obey it!</span>")
+			var/objective = "Convert as many people as possible into a [initial(desired_race.name)]. Racewar!"
+			brainwash(C, objective)
+			log_game("[key_name(C)] has been brainwashed with the objective '[objective]' via the species converter.")
+
+	iterations++
+	use_power(500)
+
+/obj/machinery/species_converter/proc/start_extracting()
```
Comment (park66665) [inline code/game/machinery/race_converter.dm]: I do think we need a better name than copypasta of `/obj/machinery/fat_sucker/proc/start_extracting()`

**Review by park66665 — COMMENTED** [review]
Function-wise, the chamber works as intended, This is mostly about code quality. The code won't be squeaky clean nor crystal clear (but hey I bet less than 1000 lines of code in SS13 currently are) but I am sure that after these are addressed, this PR can be good to go.

File: code/game/machinery/race_converter.dm (hunk 8 above)
Comment (utilbird) [reply code/game/machinery/race_converter.dm]: Forgot to change, oops

**Review by park66665 — APPROVED** [review]
All clear. This PR is good to go.

**Pull request merged**